pkg/server/testing/fake_execution_client.go

```diff
@@ -0,0 +1,313 @@
+	if ok {
+		return nil, containerd.ErrContainerExists
+	}
+	pid := generatePid()
+	f.ContainerList[createOpts.ID] = container.Container{
+		ID:     createOpts.ID,
+		Pid:    pid,
+		Status: container.Status_CREATED,
+	}
+	f.sendEvent(&container.Event{
+		ID:   createOpts.ID,
+		Type: container.Event_CREATE,
+		Pid:  pid,
+	})
+	return &execution.CreateResponse{
+		ID:  createOpts.ID,
+		Pid: pid,
+	}, nil
+}
+
+// Start is a test implementation of execution.Start
+func (f *FakeExecutionClient) Start(ctx context.Context, startOpts *execution.StartRequest, opts ...grpc.CallOption) (*google_protobuf.Empty, error) {
+	f.Lock()
+	defer f.Unlock()
+	f.appendCalled("start", startOpts)
+	if err := f.popError("start"); err != nil {
+		return nil, err
+	}
+	c, ok := f.ContainerList[startOpts.ID]
+	if !ok {
+		return nil, containerd.ErrContainerNotExist
+	}
+	f.sendEvent(&container.Event{
+		ID:   c.ID,
+		Type: container.Event_START,
+		Pid:  c.Pid,
+	})
+	switch c.Status {
+	case container.Status_CREATED:
+		c.Status = container.Status_RUNNING
+		return &google_protobuf.Empty{}, nil
+	case container.Status_STOPPED:
+		return &google_protobuf.Empty{}, fmt.Errorf("cannot start a container that has stopped")
+	case container.Status_RUNNING:
+		return &google_protobuf.Empty{}, fmt.Errorf("cannot start an already running container")
+	default:
+		return &google_protobuf.Empty{}, fmt.Errorf("cannot start a container in the %s state", c.Status)
+	}
+}
+
+// Delete is a test implementation of execution.Delete
+func (f *FakeExecutionClient) Delete(ctx context.Context, deleteOpts *execution.DeleteRequest, opts ...grpc.CallOption) (*execution.DeleteResponse, error) {
+	f.Lock()
+	defer f.Unlock()
+	f.appendCalled("delete", deleteOpts)
+	if err := f.popError("delete"); err != nil {
+		return nil, err
+	}
+	c, ok := f.ContainerList[deleteOpts.ID]
+	if !ok {
+		return nil, containerd.ErrContainerNotExist
+	}
+	delete(f.ContainerList, deleteOpts.ID)
+	f.sendEvent(&container.Event{
+		ID:   c.ID,
+		Type: container.Event_EXIT,
+		Pid:  c.Pid,
+	})
+	return nil, nil
+}
+
+// Info is a test implementation of execution.Info
+func (f *FakeExecutionClient) Info(ctx context.Context, infoOpts *execution.InfoRequest, opts ...grpc.CallOption) (*container.Container, error) {
+	f.Lock()
+	defer f.Unlock()
+	f.appendCalled("info", infoOpts)
+	if err := f.popError("info"); err != nil {
+		return nil, err
+	}
+	c, ok := f.ContainerList[infoOpts.ID]
+	if !ok {
+		return nil, containerd.ErrContainerNotExist
+	}
+	return &c, nil
+}
+
+// List is a test implementation of execution.List
+func (f *FakeExecutionClient) List(ctx context.Context, listOpts *execution.ListRequest, opts ...grpc.CallOption) (*execution.ListResponse, error) {
+	f.Lock()
+	defer f.Unlock()
+	f.appendCalled("list", listOpts)
+	if err := f.popError("list"); err != nil {
+		return nil, err
+	}
+	resp := &execution.ListResponse{}
+	for _, c := range f.ContainerList {
+		resp.Containers = append(resp.Containers, &c)
```

Comment:
You may not want to get a pointer of the iterator.

Try this https://play.golang.org/p/6Au4lw5Zi6